openage/convert/opus/opusenc.pyx

```diff
@@ -0,0 +1,415 @@
+          * a faster version?
+          * a better interpolation?
+    '''
+    cdef:
+        int wframe_len = opts['wframe_len']
+        int ch, channel = opts['channels']
+        int bit_depth = opts['bits_per_sample']
+        opus.opus_int16 mask = ~((1 << (16 - bit_depth)) - 1)
+        float alpha
+
+        size_t iin = 0, iout = 0
+        size_t ismpls, osmpls  # number of input and output samples
+        short a,b
+        int num, den, divisor
+
+        size_t outsize
+        char *out
+
+    ismpls = inp_len // wframe_len
+    if ismpls < 2:
+        # Nothing to do.
+        ret = bytearray(inp_len)
+        memcpy(PyByteArray_AS_STRING(ret), inp, inp_len)
+        return ret
+
+    divisor = gcd(target_rate, opts['input_rate'])
+    num = <int> (target_rate / divisor)
+    den = <int> (opts['input_rate'] / divisor)
+
+    # We're using linear interpolation. That means, we need for every output
+    # sample one input samples that comes before and one after the output
+    # sample. ismpls - 1 is the index of the last input sample. By multiplying
+    # with num / den we get the index of the last output sample. In general,
+    # this is not a whole number. Therefore we take only the integer part.
+    # Because indices are zero-based, add one for the number of output samples.
+    osmpls = ((ismpls - 1) * num / den).__trunc__() + 1
+    outsize = osmpls * channel * 2  # we convert to int16
+
+    ret = bytearray(outsize)
+    out = PyByteArray_AS_STRING(ret)
+
+    # iout is the next sample we want to calculate. We need one sample earlier
+    # and one later in time. Those are at indices iin and iin + 1. We need to
+    # check in every iteration, if iout is now after iin + 1 and in that case
+    # increase iin by one. Because we are _up_-sampling, the time between two
+    # output samples is shorter than between two input samples, and therefore
+    # adding one to iin suffices to keep the invariant
+    #           iin * num/den ≤ iout ≤ (iin + 1) * num/den .
+    if bit_depth > 8:
+        for iout in range(0, osmpls):
+            if (iin + 1) * num < iout * den:
+                iin += 1
+
+            alpha = (<float> iout * den) / num - iin
+            for ch in range(0, channel):
+                a = (<short *> inp)[iin * channel + ch]
+                b = (<short *> inp)[(iin + 1) * channel + ch]
+                (<opus.opus_int16 *> out)[iout * channel + ch] = \
+                        (<opus.opus_int16> (a + alpha * (b - a))) & mask
+    else:
+        for iout in range(0, osmpls):
+            if (iin + 1) * num < iout * den:
+                iin += 1
+
+            alpha = (<float> iout * den) / num - iin
+            for ch in range(0, channel):
+                a = (<char *> inp)[iin * channel + ch]
+                b = (<char *> inp)[(iin + 1) * channel + ch]
+                (<opus.opus_int16 *> out)[iout * channel + ch] = \
+                        (<opus.opus_int16> (a + alpha * (b - a))) << 8 & mask
```

Comment:
the only difference is the `<< 8`, right? we could init that number to 0 or 8 depending what bit_depth is.

```python
if bit_deptn > 8:
    shift_correct = 0
else:
    shift_correct = 8
```

and always do `<< shift_correct`.

Reply:
and the casts are to `char *` instead of `short *`

Reply:
Oh, didn't see that, sorry. Can you add a comment in the code what differences are needed and why?

Reply:
done